components/script/dom/messageport.rs

```diff
@@ -0,0 +1,523 @@
+            self.outgoing_message_buffer.borrow_mut().push_back(task);
+            return;
+        }
+
+        let target_port_id = match *self.entangled_port.borrow() {
+            Some(port_id) => port_id.clone(),
+            None => return,
+        };
+
+        if self.has_been_shipped.get() {
+            if let Some(sender) = &*self.entangled_sender.borrow() {
+                let _ = sender.send(MessagePortMsg::NewTask(target_port_id, task));
+            } else {
+                // Note: this is the mirror case of when we're awaiting transfer.
+                //
+                // In case the entangled port has been shipped, but we haven't received the new sender yet.
+                // This could happen if a port is shipped in a task because it's entangled port is transferred,
+                // and the same task immediately starts sending messages meant for the transferred port.
+                self.outgoing_message_buffer.borrow_mut().push_back(task);
+            }
+        } else {
+            let this = Trusted::new(&*self.global());
+            let _ = self.global().port_message_queue().queue(
+                task!(process_pending_port_messages: move || {
+                    let global = this.root();
+                    global.upcast::<GlobalScope>().route_task_to_port(target_port_id, task);
+                }),
+                &self.global(),
+            );
+        }
+    }
+
+    #[allow(unsafe_code)]
+    fn post_message_impl(
+        &self,
+        cx: *mut JSContext,
+        message: HandleValue,
+        transfer: Vec<*mut JSObject>,
+    ) -> ErrorResult {
+        if self.detached.get() || self.transferred.get() {
+            return Ok(());
+        }
+        // Step 1
+        let target_port = self.entangled_port.borrow();
+
+        // Step 3
+        let mut doomed = false;
+
+        rooted!(in(cx) let mut val = UndefinedValue());
+        let ports = transfer
+            .iter()
+            .filter_map(|&obj| root_from_object::<MessagePort>(obj, cx).ok());
+        for port in ports {
+            // Step 2
+            if port.message_port_id() == self.message_port_id() {
+                return Err(Error::DataClone);
+            }
+
+            // Step 4
+            if let Some(target_id) = target_port.as_ref() {
+                if port.message_port_id() == target_id {
+                    doomed = true;
+                }
+            }
+        }
+
+        unsafe { transfer.to_jsval(cx, val.handle_mut()) };
+
+        // Step 5
+        let data = StructuredCloneData::write(cx, message, val.handle())?.move_to_arraybuffer();
+
+        if doomed {
+            return Ok(());
+        }
+
+        // Step 6
+        if target_port.is_none() {
+            // TODO: find a way to deal with target_port being none
+            // because we haven't complete the transfer.
+            // One way would be to transfer the entangled port ID along with our own ID.
```

Comment:
Note: I'm not sure whether this is a problem in practice. When both a port and it's entangled port are transferred together, and then are received in the same task, then within that task, they are both in "awaiting completion of transfer" mode and don't know which port they're entangled with. 

So yes, one could do a `postMessage` call and transfer the entangled. However that message will never be received, since initially it will be buffered until the transfer completes, and after that the entanlged port will be marked as "transferred", so it won't handle incoming messages. 

So the message is effectively "doomed", however we won't know until later, hence we won't be able to "optionally report to a developer console that the target port was posted to itself, causing the communication channel to be lost" as the spec suggests.

Reply:
We could either find a way to transfer the entangled Id as well, I tried using the `content: *mut *mut raw::c_void` at some point and I was getting different IDs on the other side, although at that time I just assigned the `MessagePortId` as a whole, I didn't try using a pointer to a `u64` instead. 

Currently the `extra_data: *mut u64` is used to transfer the Id of the port itself, and there is no room in that for another Id. `extra_data` has proven a reliable means to transfer a single ID across. 

If we absolutely must know the entangled ID at this point, we could also do a blocking call to the constellation and get it.